pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,154 @@
+package shortnames
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/pkg/sysregistriesv2"
+	"github.com/containers/image/v5/types"
+	"github.com/manifoldco/promptui"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+// The default registry to use in case no unqualified-search registries are
+// specified.
+const defaultRegistry = "docker.io"
+
+// IsShortName returns true if the specified input is a "short name" and an
+// error that may occur when the input could not be parsed.  A "short name"
+// refers to a container image without a fully-qualified reference, and is
+// hence missing a registry (or domain).
+//
+// Examples:
+//  * short names: "image:tag", "library/fedora"
+//  * not short names: "quay.io/image", "localhost/image:tag", "server.org:5000/lib/image"
+func IsShortName(input string) (bool, error) {
+	repo, err := reference.Parse(input)
+	if err != nil {
+		return false, errors.Wrapf(err, "cannot parse input: %q", input)
+	}
+	named, ok := repo.(reference.Named)
+	if !ok {
+		return false, nil
+	}
+
+	registry := reference.Domain(named)
+	if strings.ContainsAny(registry, ".:") || registry == "localhost" {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func completeNameWithRegistry(name, registry string) string {
+	return fmt.Sprintf("%s/%s", registry, name)
+}
+
+// Resolve resolves the specified name to either one or more fully-qualified
+// image references.  If the specified name is already a fully-qualified
+// reference (i.e., not a short name), it is returned as is.  In case, it's
+// a short name, it's resolved as follows:
+//
+// If there is a matching short-name alias in the registries.conf, the alias
+// is returned.  Otherwise, the behaviour is determined by the specified
+// short-name mode as configured in the registries.conf:
+//
+// * "" or "disabled": append the name to all specified unqualified-search
+//                     registries (USR).
+//
+// * "enforcing": prompt the user to select a USR.  Error if stdout is not a
+//                TTY as we cannot prompt.
+//
+// * "permissive": prompt the user to select a USR.  If stdout is not a TTY,
+//                 fall back to the "disabled" behaviour.
+//
+// Note that if there is only one USR, the USR is auto-selected without prompt.
+// If there is no USR, "docker.io" is selected.
+//
+// If `clear` is set, a possibly existing alias will be cleared.
+func Resolve(ctx *types.SystemContext, name string, clear bool) ([]string, error) {
+	// If `input` is not a short name, we're done.
+	isShort, err := IsShortName(name)
```

Comment:
(This might very well be fine, mostly a note to self: I’m wondering whether this belongs here, vs. other heuristics like “is this an image ID prefix”? I’ll need to check the uses)